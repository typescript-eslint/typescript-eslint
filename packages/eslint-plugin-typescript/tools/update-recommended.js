"use strict";

const path = require("path");
const fs = require("fs");
const requireIndex = require("requireindex");

/**
 * Generate recommended configuration
 * @returns {void}
 */
function generate() {
    // replace this with Object.entries when node > 8
    const allRules = requireIndex(path.resolve(__dirname, "../lib/rules"));
    const rules = Object.keys(allRules)
        .filter(key => allRules[key].meta.docs.recommended)
        .reduce((config, item) => {
            config[item] = "error";
            return config;
        }, {});

    const filePath = path.resolve(__dirname, "../lib/configs/recommended.json");

    fs.writeFileSync(filePath, `${JSON.stringify({ rules }, null, 4)}\n`);
}

generate();
