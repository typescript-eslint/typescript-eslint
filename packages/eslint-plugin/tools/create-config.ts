/* eslint-disable no-console */

import path from 'path';
import fs from 'fs';
import requireIndex from 'requireindex';

const MAX_RULE_NAME_LENGTH = 33 + 'typescript-eslint/'.length;
const RULE_NAME_PREFIX = '@typescript-eslint';

const all = Object.entries(
  requireIndex(path.resolve(__dirname, '../dist/rules')),
);

const baseConfig = {
  parser: '@typescript-eslint/parser',
  parserOptions: {
    sourceType: 'module',
  },
  plugins: ['@typescript-eslint'],
};

const bannedRules = new Set([
  'camelcase',
  'indent',
  'no-array-constructor',
  'no-unused-vars',
]);

/**
 * Helper function reduces records to key - value pairs.
 * @param config
 * @param entry
 */
const reducer = (
  config: Record<string, string>,
  entry: Record<string, any>,
): Record<string, string> => {
  const key = entry[0];
  const value = entry[1];
  const ruleName = `${RULE_NAME_PREFIX}/${key}`;
  const setting = value.default.meta.docs.recommended;
  const usedSetting = !setting ? 'warn' : setting;

  // having this here is just for output niceness (the keys will be ordered)
  if (bannedRules.has(key)) {
    console.log(key.padEnd(MAX_RULE_NAME_LENGTH), '= off');
    config[key] = 'off';
  }
  console.log(ruleName.padEnd(MAX_RULE_NAME_LENGTH), '=', usedSetting);
  config[ruleName] = usedSetting;

  return config;
};

/**
 * Helper function to check for invalid recommended setting.
 */
function checkValidSettings(): boolean {
  const validSettings = ['error', 'warn', false];
  let result = true;

  all.forEach(entry => {
    const key = entry[0];
    const value = entry[1];
    const setting = value.default.meta.docs.recommended;

    if (!validSettings.includes(setting)) {
      console.error(`ERR! Invalid level for rule ${key}: "${setting}"`);
      result = false;
    }
  });

  return result;
}

/**
 * Helper function generates configuration.
 */
function generate(rules: Record<string, string>, filePath: string): void {
  const config = Object.assign({}, baseConfig, { rules });

  fs.writeFileSync(filePath, `${JSON.stringify(config, null, 2)}\n`);
}

if (checkValidSettings()) {
  console.log('------------------------- all.json -------------------------');
  generate(
    all.reduce(reducer, {}),
    path.resolve(__dirname, '../src/configs/all.json'),
  );

  console.log('--------------------- recommended.json ---------------------');
  const recommendedSettings = ['error', 'warn'];
  generate(
    all
      .filter(entry =>
        recommendedSettings.includes(entry[1].default.meta.docs.recommended),
      )
      .reduce(reducer, {}),
    path.resolve(__dirname, '../src/configs/recommended.json'),
  );
}
