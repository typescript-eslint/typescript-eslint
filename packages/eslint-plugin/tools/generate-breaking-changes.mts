import type { TypeScriptESLintRules } from '@typescript-eslint/eslint-plugin/use-at-your-own-risk/rules';
import { fetch } from 'cross-fetch';
// markdown-table is ESM, hence this file needs to be `.mts`
import { markdownTable } from 'markdown-table';

async function main(): Promise<void> {
  const rulesImport = await import('../src/rules/index.js');
  /*
  weird TS resolution which adds an additional default layer in the type like:
      { default: { default: Rules }}
  instead of just
      { default: Rules }
  @ts-expect-error */
  const rules = rulesImport.default as TypeScriptESLintRules;

  // Annotate which rules are new since the last version
  async function getNewRulesAsOfMajorVersion(
    oldVersion: string,
  ): Promise<Set<string>> {
    // 1. Get the current list of rules (already done)
    const newRuleNames = Object.keys(rules);

    // 2. Retrieve the old version of typescript-eslint from unpkg
    const oldUrl = `https://unpkg.com/@typescript-eslint/eslint-plugin@${oldVersion}/dist/configs/all.js`;
    const oldFileText = await (await fetch(oldUrl)).text();
    const oldObjectText = oldFileText.substring(
      oldFileText.indexOf('{'),
      oldFileText.lastIndexOf('}') + 1,
    );
    // Normally we wouldn't condone using the 'eval' API...
    // But this is an internal-only script and it's the easiest way to convert
    // the JS raw text into a runtime object. 🤷
    let oldRulesObject!: { rules: TypeScriptESLintRules };
    eval('oldRulesObject = ' + oldObjectText);
    const oldRuleNames = new Set(Object.keys(oldRulesObject.rules));

    // 3. Get the keys that exist in (1) (new version) and not (2) (old version)
    return new Set(
      newRuleNames.filter(
        newRuleName => !oldRuleNames.has(`@typescript-eslint/${newRuleName}`),
      ),
    );
  }

  const newRuleNames = await getNewRulesAsOfMajorVersion('5.0.0');

  console.log(`## Table Key

<table>
  <thead>
    <tr>
      <th>Column</th>
      <th>Description</th>
      <th>Emojis</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>Status</td>
      <td>Being added, deprecated, or removed</td>
      <td>
        <ul>
          <li>🆕 = newly added to typescript-eslint</li>
          <li>🙅 = to be deprecated in the next major</li>
          <li>💀 = currently deprecated; to be removed in the next version</li>
        </ul>
      </td>
    </tr>
    <tr>
      <td>TC</td>
      <td>Requires type checking?</td>
      <td>
        <ul>
          <li>💭 = yes</li>
        </ul>
      </td>
    </tr>
    <tr>
      <td>Ext</td>
      <td>Extension rule?</td>
      <td>
        <ul>
          <li>🧱 = yes</li>
        </ul>
      </td>
    </tr>
    <tr>
      <td>Rec'd</td>
      <td>Recommended</td>
      <td>
        <ul>
          <li>➕ = add to recommended this version</li>
          <li>➖️ = remove from recommended this version</li>
          <li>🟩 = stays in recommended this version</li>
        </ul>
      </td>
    </tr>
    <tr>
      <td>Strict</td>
      <td>Strict</td>
      <td>
        <ul>
        <li>➕ = add to strict this version</li>
        <li>➖️ = remove from strict this version</li>
        <li>🔵 = stays in strict this version</li>
        </ul>
      </td>
    </tr>
    <tr>
      <td>Style</td>
      <td>Style</td>
      <td>
        <ul>
          <li>➕ = add to stylistic this version</li>
          <li>➖️ = remove from stylistic this version</li>
          <li>🔸 = stays in stylistic this version</li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>

## Recommendations Table

> Hint: search for 🆕 to find newly added rules, and ➕ or ➖ to see config changes.
`);

  console.log(
    markdownTable([
      ['Rule', 'Status', 'TC', 'Ext', "Rec'd", 'Strict', 'Style', 'Comment'],
      ...Object.entries(rules).map(([ruleName, { meta }]) => {
        const { deprecated } = meta;
        const { extendsBaseRule, recommended, requiresTypeChecking } =
          meta.docs!;

        return [
          `[\`${ruleName}\`](https://typescript-eslint.io/rules/${ruleName})`,
          newRuleNames.has(ruleName) ? '🆕' : deprecated ? '💀' : '',
          requiresTypeChecking ? '💭' : '',
          extendsBaseRule ? '🧱' : '',
          recommended === 'recommended' ? '🟩' : '',
          recommended === 'strict' ? '🔵' : '',
          recommended === 'stylistic' ? '🔸' : '',
          meta.type === 'layout' ? 'layout 📐' : '(todo)',
        ];
      }),
    ]),
  );
}

main().catch(error => {
  console.error(error);
});
