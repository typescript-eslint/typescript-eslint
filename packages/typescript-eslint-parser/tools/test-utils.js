/**
 * @fileoverview Tools for running test cases
 * @author Nicholas C. Zakas
 * @copyright jQuery Foundation and other contributors, https://jquery.org/
 * MIT License
 */

"use strict";

//------------------------------------------------------------------------------
//   Private
//--------------------------------------------------------------------------------

/**
 * Returns a raw copy of the given AST
 * @param  {Object} ast the AST object
 * @returns {Object}     copy of the AST object
 */
function getRaw(ast) {
    return JSON.parse(JSON.stringify(ast, (key, value) => {
        if ((key === "start" || key === "end") && typeof value === "number") {
            return undefined;
        }

        return value;
    }));
}

module.exports = {
    getRaw
};
