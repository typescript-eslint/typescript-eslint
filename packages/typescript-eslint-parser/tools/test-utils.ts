import * as parser from '../src/parser';

/**
 * Returns a raw copy of the given AST
 * @param  {Object} ast the AST object
 * @returns {Object}     copy of the AST object
 */
function getRaw(ast: any) {
  return JSON.parse(
    JSON.stringify(ast, (key, value) => {
      if ((key === 'start' || key === 'end') && typeof value === 'number') {
        return undefined;
      }
      return value;
    })
  );
}

/**
 * Returns a function which can be used as the callback of a Jest test() block,
 * and which performs an assertion on the snapshot for the given code and config.
 * @param {string} code The source code to parse
 * @param {*} config the parser configuration
 * @returns {Function} callback for Jest test() block
 */
export function createSnapshotTestBlock(code: any, config = {}) {
  const defaultConfig = {
    loc: true,
    range: true,
    raw: true,
    tokens: true,
    comment: true,
    errorOnUnknownASTType: true,
    sourceType: 'module'
  };
  config = Object.assign({}, defaultConfig, config);

  /**
   * @returns {Object} the AST object
   */
  function parse() {
    const ast = parser.parseForESLint(code, config).ast;
    return getRaw(ast);
  }

  return () => {
    try {
      const result = parse();
      expect(result).toMatchSnapshot();
    } catch (e) {
      /**
       * If we are deliberately throwing because of encountering an unknown
       * AST_NODE_TYPE, we rethrow to cause the test to fail
       */
      if (e.message.match('Unknown AST_NODE_TYPE')) {
        throw new Error(e);
      }
      expect(parse).toThrowErrorMatchingSnapshot();
    }
  };
}

export function formatSnapshotName(
  filename: any,
  fixturesDir: any,
  fileExtension = '.js'
) {
  return `fixtures/${filename
    .replace(fixturesDir + '/', '')
    .replace(fileExtension, '')}`;
}
