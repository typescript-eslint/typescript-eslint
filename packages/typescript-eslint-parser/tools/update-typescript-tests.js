/**
 * @fileoverview A simple script to update existing tests to reflect new
 *      parser changes.
 * @author Nicholas C. Zakas

"use strict";

/*
 * Usage:
 *      node tools/update-typescript-tests.js
 *
 */

//------------------------------------------------------------------------------
// Requirements
//------------------------------------------------------------------------------

var shelljs = require("shelljs"),
    parser = require("../parser"),
    tester = require("../tests/lib/tester"),
    path = require("path");

//------------------------------------------------------------------------------
// Helpers
//------------------------------------------------------------------------------

/**
 * Returns a raw copy of the given AST
 * @param  {object} ast the AST object
 * @returns {object}     copy of the AST object
 */
function getRaw(ast) {
    return JSON.parse(JSON.stringify(ast));
}


function getExpectedResult(code, config) {
    try {
        return tester.getRaw(parser.parse(code, config));
    } catch (ex) {
        var raw = getRaw(ex);
        raw.message = ex.message;
        return raw;
    }
}

function getTestFilenames(directory) {
    return shelljs.find(directory).filter(function(filename) {
        return filename.indexOf(".src.ts") > -1;
    }).map(function(filename) {
        return filename.substring(directory.length - 1, filename.length - 7);  // strip off ".src.ts"
    });
}

function outputResult(result, testResultFilename) {
    shelljs.echo("module.exports = " + JSON.stringify(result, null, "    ") + ";").to(testResultFilename);
}

//------------------------------------------------------------------------------
// Setup
//------------------------------------------------------------------------------

var FIXTURES_DIR = "./tests/fixtures/typescript";

var testFiles = getTestFilenames(FIXTURES_DIR);


// update all tests in ecma-features
testFiles.forEach(function(filename) {

    var feature = path.dirname(filename),
        code = shelljs.cat(path.resolve(FIXTURES_DIR, filename) + ".src.ts"),
        config = {
            loc: true,
            range: true,
            tokens: true,
            ecmaFeatures: {},
            errorOnUnknownASTType: true
        };

    var testResultFilename = path.resolve(__dirname, "..", FIXTURES_DIR, filename) + ".result.js";
    var result = getExpectedResult(code, config);

    outputResult(result, testResultFilename);
});
