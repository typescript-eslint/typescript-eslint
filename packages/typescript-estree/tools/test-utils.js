/**
 * @fileoverview Tools for running test cases
 * @author Nicholas C. Zakas
 * @author James Henry <https://github.com/JamesHenry>
 * @copyright jQuery Foundation and other contributors, https://jquery.org/
 * MIT License
 */

'use strict';

//------------------------------------------------------------------------------
// Requirements
//------------------------------------------------------------------------------

const parser = require('../parser');

//------------------------------------------------------------------------------
//   Private
//--------------------------------------------------------------------------------

/**
 * Returns a raw copy of the given AST
 * @param  {Object} ast the AST object
 * @returns {Object}     copy of the AST object
 */
function getRaw(ast) {
  return JSON.parse(
    JSON.stringify(ast, (key, value) => {
      if ((key === 'start' || key === 'end') && typeof value === 'number') {
        return undefined;
      }
      return value;
    })
  );
}

/**
 * Returns a function which can be used as the callback of a Jest test() block,
 * and which performs an assertion on the snapshot for the given code and config.
 * @param {string} code The source code to parse
 * @param {*} config the parser configuration
 * @returns {Function} callback for Jest test() block
 */
function createSnapshotTestBlock(code, config) {
  /**
   * @returns {Object} the AST object
   */
  function parse() {
    const ast = parser.parse(code, config);
    return getRaw(ast);
  }

  return () => {
    try {
      const result = parse();
      expect(result).toMatchSnapshot();
    } catch (e) {
      /**
       * If we are deliberately throwing because of encountering an unknown
       * AST_NODE_TYPE, we rethrow to cause the test to fail
       */
      if (e.message.match('Unknown AST_NODE_TYPE')) {
        throw new Error(e);
      }
      expect(parse).toThrowErrorMatchingSnapshot();
    }
  };
}

module.exports = {
  getRaw,
  createSnapshotTestBlock
};
