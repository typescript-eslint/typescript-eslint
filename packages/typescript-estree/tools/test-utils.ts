import * as parser from '../src/parser';
import { ParserOptions } from '../src/parser-options';

/**
 * Returns a raw copy of the given AST
 * @param  {Object} ast the AST object
 * @returns {Object}     copy of the AST object
 */
export function getRaw(ast: any) {
  return JSON.parse(
    JSON.stringify(ast, (key, value) => {
      if ((key === 'start' || key === 'end') && typeof value === 'number') {
        return undefined;
      }
      return value;
    }),
  );
}

export function parseCodeAndGenerateServices(
  code: string,
  config: ParserOptions,
) {
  return parser.parseAndGenerateServices(code, config);
}

/**
 * Returns a function which can be used as the callback of a Jest test() block,
 * and which performs an assertion on the snapshot for the given code and config.
 * @param {string} code The source code to parse
 * @param {ParserOptions} config the parser configuration
 * @param {boolean} generateServices Flag determining whether to generate ast maps and program or not
 * @returns {jest.ProvidesCallback} callback for Jest it() block
 */
export function createSnapshotTestBlock(
  code: string,
  config: ParserOptions,
  generateServices?: true,
) {
  /**
   * @returns {Object} the AST object
   */
  function parse() {
    const ast = generateServices
      ? parser.parseAndGenerateServices(code, config).ast
      : parser.parse(code, config);
    return getRaw(ast);
  }

  return () => {
    try {
      const result = parse();
      expect(result).toMatchSnapshot();
    } catch (e) {
      /**
       * If we are deliberately throwing because of encountering an unknown
       * AST_NODE_TYPE, we rethrow to cause the test to fail
       */
      if (e.message.match('Unknown AST_NODE_TYPE')) {
        throw new Error(e);
      }
      expect(parse).toThrowErrorMatchingSnapshot();
    }
  };
}

export function formatSnapshotName(
  filename: string,
  fixturesDir: string,
  fileExtension = '.js',
): string {
  return `fixtures/${filename
    .replace(fixturesDir + '/', '')
    .replace(fileExtension, '')}`;
}

/**
 * Check if file extension is one used for jsx
 * @param fileType
 */
export function isJSXFileType(fileType: string): boolean {
  if (fileType.startsWith('.')) {
    fileType = fileType.slice(1);
  }
  return fileType === 'js' || fileType === 'jsx' || fileType === 'tsx';
}
