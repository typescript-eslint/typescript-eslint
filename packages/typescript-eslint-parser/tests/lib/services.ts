import path from 'path';
import fs from 'fs';
import glob from 'glob';
import * as testUtils from '../../tools/test-utils';

//------------------------------------------------------------------------------
// Setup
//------------------------------------------------------------------------------

const FIXTURES_DIR = './tests/fixtures/services';
const testFiles = glob.sync(`${FIXTURES_DIR}/**/*.src.ts`);

function createConfig(filename: string): object {
  return {
    filePath: filename,
    generateServices: true,
    project: './tsconfig.json',
    tsconfigRootDir: path.resolve(FIXTURES_DIR)
  };
}

//------------------------------------------------------------------------------
// Tests
//------------------------------------------------------------------------------

describe('services', () => {
  testFiles.forEach(filename => {
    const code = fs.readFileSync(filename, 'utf8');
    const config = createConfig(filename);
    it(
      testUtils.formatSnapshotName(filename, FIXTURES_DIR, '.ts'),
      testUtils.createSnapshotTestBlock(code, config)
    );
    it(`${testUtils.formatSnapshotName(
      filename,
      FIXTURES_DIR,
      '.ts'
    )} services`, () => {
      testUtils.testServices(code, config);
    });
  });
});
