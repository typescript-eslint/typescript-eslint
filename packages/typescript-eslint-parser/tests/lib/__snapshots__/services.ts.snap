// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`services fixtures/isolated-file.src 1`] = `
Object {
  "body": Array [
    Object {
      "declarations": Array [
        Object {
          "id": Object {
            "loc": Object {
              "end": Object {
                "column": 7,
                "line": 1,
              },
              "start": Object {
                "column": 6,
                "line": 1,
              },
            },
            "name": "x",
            "range": Array [
              6,
              7,
            ],
            "type": "Identifier",
          },
          "init": Object {
            "elements": Array [
              Object {
                "loc": Object {
                  "end": Object {
                    "column": 12,
                    "line": 1,
                  },
                  "start": Object {
                    "column": 11,
                    "line": 1,
                  },
                },
                "range": Array [
                  11,
                  12,
                ],
                "raw": "3",
                "type": "Literal",
                "value": 3,
              },
              Object {
                "loc": Object {
                  "end": Object {
                    "column": 15,
                    "line": 1,
                  },
                  "start": Object {
                    "column": 14,
                    "line": 1,
                  },
                },
                "range": Array [
                  14,
                  15,
                ],
                "raw": "4",
                "type": "Literal",
                "value": 4,
              },
              Object {
                "loc": Object {
                  "end": Object {
                    "column": 18,
                    "line": 1,
                  },
                  "start": Object {
                    "column": 17,
                    "line": 1,
                  },
                },
                "range": Array [
                  17,
                  18,
                ],
                "raw": "5",
                "type": "Literal",
                "value": 5,
              },
            ],
            "loc": Object {
              "end": Object {
                "column": 19,
                "line": 1,
              },
              "start": Object {
                "column": 10,
                "line": 1,
              },
            },
            "range": Array [
              10,
              19,
            ],
            "type": "ArrayExpression",
          },
          "loc": Object {
            "end": Object {
              "column": 19,
              "line": 1,
            },
            "start": Object {
              "column": 6,
              "line": 1,
            },
          },
          "range": Array [
            6,
            19,
          ],
          "type": "VariableDeclarator",
        },
      ],
      "kind": "const",
      "loc": Object {
        "end": Object {
          "column": 20,
          "line": 1,
        },
        "start": Object {
          "column": 0,
          "line": 1,
        },
      },
      "range": Array [
        0,
        20,
      ],
      "type": "VariableDeclaration",
    },
  ],
  "comments": Array [],
  "loc": Object {
    "end": Object {
      "column": 0,
      "line": 2,
    },
    "start": Object {
      "column": 0,
      "line": 1,
    },
  },
  "range": Array [
    0,
    21,
  ],
  "sourceType": "module",
  "tokens": Array [
    Object {
      "loc": Object {
        "end": Object {
          "column": 5,
          "line": 1,
        },
        "start": Object {
          "column": 0,
          "line": 1,
        },
      },
      "range": Array [
        0,
        5,
      ],
      "type": "Keyword",
      "value": "const",
    },
    Object {
      "loc": Object {
        "end": Object {
          "column": 7,
          "line": 1,
        },
        "start": Object {
          "column": 6,
          "line": 1,
        },
      },
      "range": Array [
        6,
        7,
      ],
      "type": "Identifier",
      "value": "x",
    },
    Object {
      "loc": Object {
        "end": Object {
          "column": 9,
          "line": 1,
        },
        "start": Object {
          "column": 8,
          "line": 1,
        },
      },
      "range": Array [
        8,
        9,
      ],
      "type": "Punctuator",
      "value": "=",
    },
    Object {
      "loc": Object {
        "end": Object {
          "column": 11,
          "line": 1,
        },
        "start": Object {
          "column": 10,
          "line": 1,
        },
      },
      "range": Array [
        10,
        11,
      ],
      "type": "Punctuator",
      "value": "[",
    },
    Object {
      "loc": Object {
        "end": Object {
          "column": 12,
          "line": 1,
        },
        "start": Object {
          "column": 11,
          "line": 1,
        },
      },
      "range": Array [
        11,
        12,
      ],
      "type": "Numeric",
      "value": "3",
    },
    Object {
      "loc": Object {
        "end": Object {
          "column": 13,
          "line": 1,
        },
        "start": Object {
          "column": 12,
          "line": 1,
        },
      },
      "range": Array [
        12,
        13,
      ],
      "type": "Punctuator",
      "value": ",",
    },
    Object {
      "loc": Object {
        "end": Object {
          "column": 15,
          "line": 1,
        },
        "start": Object {
          "column": 14,
          "line": 1,
        },
      },
      "range": Array [
        14,
        15,
      ],
      "type": "Numeric",
      "value": "4",
    },
    Object {
      "loc": Object {
        "end": Object {
          "column": 16,
          "line": 1,
        },
        "start": Object {
          "column": 15,
          "line": 1,
        },
      },
      "range": Array [
        15,
        16,
      ],
      "type": "Punctuator",
      "value": ",",
    },
    Object {
      "loc": Object {
        "end": Object {
          "column": 18,
          "line": 1,
        },
        "start": Object {
          "column": 17,
          "line": 1,
        },
      },
      "range": Array [
        17,
        18,
      ],
      "type": "Numeric",
      "value": "5",
    },
    Object {
      "loc": Object {
        "end": Object {
          "column": 19,
          "line": 1,
        },
        "start": Object {
          "column": 18,
          "line": 1,
        },
      },
      "range": Array [
        18,
        19,
      ],
      "type": "Punctuator",
      "value": "]",
    },
    Object {
      "loc": Object {
        "end": Object {
          "column": 20,
          "line": 1,
        },
        "start": Object {
          "column": 19,
          "line": 1,
        },
      },
      "range": Array [
        19,
        20,
      ],
      "type": "Punctuator",
      "value": ";",
    },
  ],
  "type": "Program",
}
`;
